### ISSUE TITLE
Synchronisation

### ISSUE THREAD

**Issue by creynders** [keystonejs/grappling-hook#6]
One of the things I racked my brain on when writing the parallel stuff was how to deal with `next` and `done`. ATM the synchronisation strategy is left to the consumer, i.e. if `next` is called synchronously, it will immediately call the next middleware. This has subtle but important effects, for instance:

``` js
function createParallel(name) {
    return function(next, done) {
        console.log(name + ' setup');
        setTimeout(function() {
            console.log(name + ' done');
            done();
        }, 0);
        next();// <------------------- next is called after the async stuff is fired off
    };
}

instance.pre('test', createParallel("A"), createParallel("B")).callHook('pre:test');
```

``` sh
# output
A setup
B setup
A done
B done
```

However

``` js
function createParallel(name) {
    return function(next, done) {
        console.log(name + ' setup');
        next(); // <------------------- next is called before the async stuff is fired off
        setTimeout(function() {
            console.log(name + ' done');
            done();
        }, 0);
    };
}

instance.pre('test', createParallel("A"), createParallel("B")).callHook('pre:test');
```

``` sh
# output
A setup
B setup
B done
A done
```

Now B finishes first. This has its pros and cons. 
For a long time I wanted to next-tick the parallel `next`, to be sure all of the sync code would finish before the next middleware is called.
But then I realized that it makes `next` useless. It doesn't matter where you call it in your parallel middleware, the function will execute in full, which means we might as well leave out `next` (disregarding the fact we need it to recognize parallel middleware)

On the other hand: the above might lead to some unwanted side-effects. Granted, I can't come up with a case though. I mean, if you allow your function to execute in parallel with other, possibly unknown functions, it's _your_ responsibility to make sure it _really_ doesn't matter what the other functions are doing.

Now this is just for parallel middleware, but I had to ask myself the same question regarding the "final" callback. I.e. the one that's passed to `callHook`.
ATM if all your middleware is sync, the `callHook` call is finished sync as well. I'm not entirely sure this is ok.

@jedwatson any thoughts?
Or @JohnnyEstilles ?

**Comment by creynders** [comment]
Hmmmm .... http://blog.ometer.com/2011/07/24/callbacks-synchronous-and-asynchronous/

**Comment by estilles** [comment]
@creynders, regarding the parallel hooks, I've ran into both of the scenarios you mentioned. In my use-cases, while running them in parallel I had no expectation the tasks would end in any particular order. So, while I noticed the difference, I was mostly concerned with detecting when they were all completed.

Have you ran into any use-cases in which the completion order was important?

I'll check out that article ...

**Comment by creynders** [comment]
No I haven't. I think the only situation in which it really matters is when errors occur, e.g. parallel A calls `next` before calling its async operation, but the next middleware function errors synchronously (either by throwing if it's a sync function or by immediately calling `next(err)` if it's serial/parallel).

**Comment by creynders** [comment]
I did notice there are some situations in which the "final" callback in kareem is next-ticked. AFAICT it's when
1. pre: a synchronous function is last in line
2. pre: no middleware is registered
3. post: no middleware is registered

It should be noted pre and post middleware is handled differently in kareem though. Post middleware do not receive a `next` callback, i.e. if I'm not mistaken they're always considered sync. (even though they maybe do some async stuff)

So, it's really about not finishing the hook call synchronously.

**Comment by estilles** [comment]
Your assessment/comparison with kareem is spot on. I think you're middleware is much more elegant and user-friendly. Kareem does offer the ability to call the pre/post middlewares independently, but every single one of my use-cases have followed an immediate pre/hook/post pattern, so I'm not going to miss that feature.

**Comment by estilles** [comment]
@creynders, I haven't gone through the specs yet. Do you test for erros in pre/post hooks in scenarios like the ones you mentioned above?

**Comment by creynders** [comment]
What do you mean with "test" and in which scenarios exactly? i.e. have I written tests to nail-down what should happen when for instance a synchronous function is last in line and explodes?

**Comment by creynders** [comment]
And, unless I misunderstood you with "Kareem does offer the ability to call the pre/post middlewares independently", but grappling-hook offers the same functionality through `callHook`, e.g.

``` js
grappling.callHook('pre:save');
// ~=
kareem.execPre('save');
```

**Comment by JedWatson** [comment]
> So, it's really about not finishing the hook call synchronously.

This is absolutely vital. We need to make sure that the final callback is _always_ async, and _always_ runs after every other middleware (whether serial or parallel or not async) has completed.

Otherwise we [unleash untold evil upon the world](http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony).

**Comment by creynders** [comment]
Heheh, I had a feeling a very deep rabbit hole might lie hidden there...
Fixing this ASAP

**Comment by creynders** [comment]
I think i decapitated the monster

**Comment by estilles** [comment]
> What do you mean with "test" and in which scenarios exactly?

I was referring to errors occurring during parallel calls. The question is mute now though.

> grappling-hook offers the same functionality through `callHook`

I didn't realize! `grappling-hook` is ALMOST complete! (We just need to handle `.thenable` middleware ...

**Comment by creynders** [comment]
Hmmm, I'm starting to think [this](https://github.com/keystonejs/grappling-hook/blob/master/index.js#L96) wasn't such a good idea...
It bit me in the ass right now with:

``` js
    it('should call the next middleware async even with sync parallel middleware', function(done){
        var isAsync;
        instance.hook($.PRE_TEST, function(next, done){
            isAsync = false;
            next();
            isAsync = true;
            done();
        }, function(){
            expect(false).to.be.true(); // <----- force failure
        }).callHook($.PRE_TEST, function(){
            done();
        });
    });
```

This doesn't fail since the assertion error is swallowed. Obviously, in this case I can grab it in the "final" callback, but if this confuses me, I'm pretty certain it'll confuse others too.

**Comment by creynders** [comment]
@JohnnyEstilles 

> I was referring to errors occurring during parallel calls.

Moot or not, they're tested :smile: 

> I didn't realize! grappling-hook is ALMOST complete!

It _is_ complete! Just never finished :wink:

**Comment by creynders** [comment]
Fixed in 2.2.0
